go/test/endtoend/vtorc/general/vtorc_test.go

```diff
@@ -125,112 +108,97 @@ func TestReplicaReadWrite(t *testing.T) {
+		require.NoError(t, err)
 
-	// find primary from topo
-	curPrimary := utils.ShardPrimaryTablet(t, clusterInfo, keyspace, shard0)
-	assert.NotNil(t, curPrimary, "should have elected a primary")
+		// check replication is setup correctly
+		utils.CheckReplication(t, clusterInfo, curPrimary, []*cluster.Vttablet{replica, otherReplica}, 15*time.Second)
+	})
 
-	// TODO(deepthi): we should not need to do this, the DB should be created automatically
-	_, err := curPrimary.VttabletProcess.QueryTablet(fmt.Sprintf("create database IF NOT EXISTS vt_%s", keyspace.Name), keyspace.Name, false)
-	require.NoError(t, err)
+	t.Run("ReplicationFromOtherReplica", func(t *testing.T) {
+		// point replica at otherReplica
+		changeReplicationSourceCommand := fmt.Sprintf("STOP SLAVE; RESET SLAVE ALL;"+
+			"CHANGE MASTER TO MASTER_HOST='%s', MASTER_PORT=%d, MASTER_USER='vt_repl', MASTER_AUTO_POSITION = 1; START SLAVE", utils.Hostname, otherReplica.MySQLPort)
+		_, err := utils.RunSQL(t, changeReplicationSourceCommand, replica, "")
+		require.NoError(t, err)
 
-	var replica, otherReplica *cluster.Vttablet
-	for _, tablet := range shard0.Vttablets {
-		// we know we have only two tablets, so the "other" one must be the new primary
-		if tablet.Alias != curPrimary.Alias {
-			if replica == nil {
-				replica = tablet
-			} else {
-				otherReplica = tablet
-			}
-		}
-	}
-	require.NotNil(t, replica, "should be able to find a replica")
-	require.NotNil(t, otherReplica, "should be able to find 2nd replica")
+		// wait until the source port is set back correctly by vtorc
+		utils.CheckSourcePort(t, replica, curPrimary, 15*time.Second)
 
-	// check replication is setup correctly
-	utils.CheckReplication(t, clusterInfo, curPrimary, []*cluster.Vttablet{replica, otherReplica}, 15*time.Second)
+		// check that writes succeed
+		utils.VerifyWritesSucceed(t, clusterInfo, curPrimary, []*cluster.Vttablet{replica, otherReplica}, 15*time.Second)
+	})
 
-	// point replica at otherReplica
-	changeReplicationSourceCommand := fmt.Sprintf("STOP SLAVE; RESET SLAVE ALL;"+
-		"CHANGE MASTER TO MASTER_HOST='%s', MASTER_PORT=%d, MASTER_USER='vt_repl', MASTER_AUTO_POSITION = 1; START SLAVE", utils.Hostname, otherReplica.MySQLPort)
-	_, err = utils.RunSQL(t, changeReplicationSourceCommand, replica, "")
-	require.NoError(t, err)
+	t.Run("CircularReplication", func(t *testing.T) {
+		// change the replication source on the primary
+		changeReplicationSourceCommands := fmt.Sprintf("STOP SLAVE; RESET SLAVE ALL;"+
+			"CHANGE MASTER TO MASTER_HOST='%s', MASTER_PORT=%d, MASTER_USER='vt_repl', MASTER_AUTO_POSITION = 1;"+
+			"START SLAVE;", replica.VttabletProcess.TabletHostname, replica.MySQLPort)
+		_, err := utils.RunSQL(t, changeReplicationSourceCommands, curPrimary, "")
+		require.NoError(t, err)
 
-	// wait until the source port is set back correctly by vtorc
-	utils.CheckSourcePort(t, replica, curPrimary, 15*time.Second)
+		// wait for curPrimary to reach stable state
```

Comment:
1 sec sleep .. looks to me too tight limit

Reply:
or can we not check serving status instead

Reply:
Oh that 1 second was only to wait for the previous command to finish executing. It is followed by a Wait for replication to Stop. We just don't want to be in a state when the function waiting for replication to stop executes even before replication stops.